Add PSA crypto driver delegation for AEAD single shot PSA APIs - 1

ronald-cron-arm opened the pull request:
## Description
Add PSA crypto driver delegation for AEAD single shot PSA APIs following the design defined in psa-crypto-implementation-structure.md as well as basic testing of the dispatch to driver/software fallback.

In terms of testing of the driver interface and its integration into the PSA application interface implementation, with this PR:

- all AEAD encrypt/decrypt cryptographic operations are triggered by a call to the AEAD encrypt/decrypt driver interface. The call is the same for hardware and software implementations. The AEAD driver interface is thus exercised by all the PSA unit tests that test AEAD encryption/decryption.

- the all.sh test_psa_crypto_drivers component runs all PSA unit tests with all AEAD encryptions/decryptions handled by the transparent test driver.

This PR does not add support for AEAD operations in the the all.sh test_psa_crypto_config_basic component. This will be done when #4246 is merged as additional commits in this PR or in a separate PR.

## Status
READY

## Requires Backporting
NO, PSA only.

## Steps to test or reproduce
CI only.

stevew817 commented on library/psa_crypto_aead.c:
```diff
@@ -0,0 +1,358 @@
+
+typedef struct
+{
+    const mbedtls_cipher_info_t *cipher_info;
```
Might as well clean this up while at it. cipher_info is only used in the setup call, and therefore doesn't need to be in the operation context.

stevew817 commented on library/psa_crypto_aead.c:
```diff
@@ -0,0 +1,358 @@
+        mbedtls_chachapoly_context chachapoly;
+#endif /* MBEDTLS_CHACHAPOLY_C */
+    } ctx;
+    psa_algorithm_t core_alg;
```
Might as well clean this up while at it. I don't believe it is necessary to keep track of `core_alg` in the context, since all call trees contain the algorithm.

ronald-cron-arm replied:
But we will probably need it for multi-part operation.

stevew817 replied:
I'm not going to block the PR over this, but for the record I'll say that I think that if multi-part isn't going to be implemented soon than it would still make sense to remove the unused elements until the implementation is being done. Having structs with unreferenced/unused items is a bit icky.

gilles-peskine-arm replied:
[Multipart AEAD is just around the corner](https://github.com/ARMmbed/mbedtls/pull/4338).

stevew817 replied:
Thanks, I hadn't seen that yet. I guess it deserves an explicit link to this PR?

gilles-peskine-arm replied:
Given that this PR looks almost ready but multipart AEAD is still in a draft state, I expect that @paul-elliott-arm will rebase multipart AEAD once driver delegation is merged.

stevew817 commented on library/psa_crypto_aead.c:
```diff
@@ -0,0 +1,358 @@
+#endif /* MBEDTLS_CHACHAPOLY_C */
+    } ctx;
+    psa_algorithm_t core_alg;
+    uint8_t full_tag_length;
```
Might as well clean this up while at it. `full_tag_length` is only used inside the setup call, and therefore doesn't need to be in the operation context.

stevew817 commented on library/psa_crypto_aead.c:
```diff
@@ -0,0 +1,358 @@
+    } ctx;
+    psa_algorithm_t core_alg;
+    uint8_t full_tag_length;
+    uint8_t tag_length;
```
Might as well clean this up while at it. `tag_length` is encoded in the algorithm, so it doesn't need to be stored separately.

ronald-cron-arm replied:
Ditto, we will probably need it for multi-part operations.

stevew817 commented on library/psa_crypto_aead.c:
```diff
@@ -0,0 +1,358 @@
+            mbedtls_gcm_free( &operation->ctx.gcm );
+            break;
+#endif /* MBEDTLS_GCM_C */
+    }
```
Might as well clean this up while at it, and add `mbedtls_chachapoly_free` to the switchcase (which arguably should've been there already). The only reason it doesn't fail CI is because the current chachapoly context doesn't allocate memory, so the memory leakage checkers don't understand that the free function isn't being called.

ronald-cron-arm replied:
Indeed, done.

stevew817 commented on tests/suites/test_suite_psa_crypto_driver_wrappers.data:
```diff
@@ -194,3 +194,51 @@ cipher_decrypt_multipart:PSA_ALG_CTR:PSA_KEY_TYPE_AES:"2b7e151628aed2a6abf715880
+depends_on:MBEDTLS_AES_C:MBEDTLS_GCM_C
+aead_decrypt:PSA_KEY_TYPE_AES:"a0ec7b0052541d9e9c091fb7fc481409":PSA_ALG_GCM:"00e440846db73a490573deaf3728c94f":"a3cfcb832e935eb5bc3812583b3a1b2e82920c07fda3668a35d939d8f11379bb606d39e6416b2ef336fffb15aec3f47a71e191f4ff6c56ff15913562619765b26ae094713d60bab6ab82bfc36edaaf8c7ce2cf5906554dcc5933acdb9cb42c1d24718efdc4a09256020b024b224cfe602772bd688c6c8f1041a46f7ec7d51208":"3b6de52f6e582d317f904ee7688 ...
+
+PSA AEAD decrypt, AES-GCM, 144 bytes #1, INSUFFICIENT_MEMORY
```
Why is this test case annotated with `INSUFFICIENT_MEMORY`, but expects to return `PSA_SUCCESS`?

ronald-cron-arm replied:
The return code was wrong.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -4049,107 +3897,34 @@ psa_status_t psa_aead_encrypt( mbedtls_svc_key_id_t key,
-        (void) tag;
-        return( PSA_ERROR_NOT_SUPPORTED );
-    }
+    status = psa_driver_wrapper_aead_encrypt(
```
I cherry-picked your commit "Do validation on the algorithm argument in AEAD" but changes the returned error code to PSA_ERROR_NOT_SUPPORTED as there is this in the PSA spec:
"PSA_ERROR_NOT_SUPPORTED alg is not supported or is not an AEAD algorithm."

ronald-cron-arm commented:
@stevew817 I believe I addressed all your comments and the result can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-aead-1).
Then I rebased because of the conflicts with the merge of #4246.
The resolution of the conflicts were just about replacing MBEDTLS_*_C symbols by the MBEDTLS_PSA_BUILTIN_* corresponding ones.

stevew817 commented:
Now that #4246 in in, care to add AEAD to the all.sh psa_crypto_config and psa_crypto_driver tests?

ronald-cron-arm commented:
My plan was to do that when rebasing/reworking #4240. See [comment](https://github.com/ARMmbed/mbedtls/pull/4240#issuecomment-806539963).

tom-daubney-arm reviewed (changes requested):
This PR looks mostly fine to me, however one small thing I noticed in library/psa_crypto_aead.c

The order of the conditional compilation macros is not consistent between CCM and GCM. In the `struct` declaration (Ln 33 onward), `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)` comes before `#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)`. 
The same is true in the function `psa_abort_internal` (Ln 54 onward) and the function `psa_aead_setup`.

However, in the functions `mbedtls_psa_aead_encrypt` (Ln 165 onward) and `mbedtls_psa_aead_decrypt` (Ln 275 onward), the order swaps and we have `#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)` coming before `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)`.

It would be good if this ordering was consistent and I don't think there is a good reason for them to be the way they are right now. I suggest swapping the order of the conditional compilation statements in the `mbedtls_psa_aead_encrypt` (Ln 165 onward) and `mbedtls_psa_aead_decrypt` (Ln 275 onward) functions such that the `#if defined(MBEDTLS_PSA_BUILTIN_ALG_CCM)` statements come first.

ronald-cron-arm commented:
I addressed @tom-daubney-arm comment with an additional commit and the result can be found [here](https://github.com/ronald-cron-arm/mbedtls/tree/psa-aead-2). Then I rebased onto development and resolved the conflicts: only minor conflicts due to the merge of the cipher driver delegation rework #4256.

ronald-cron-arm commented:
After the rebase from commit "psa: aead: Remove key slot from operation context" the static and inline function `psa_get_key_slot_bits()` in psa_crypto.c was not used anymore and CI jobs failing. I've updated the commit aforementioned with the removal of `psa_get_key_slot_bits()`.

stevew817 reviewed (approved):
This is good to go in IMO.

tom-daubney-arm reviewed (approved):
Thanks for the changes. LGTM.
